subiquity/server/runner.py

```diff
@@ -24,22 +24,27 @@ class LoggedCommandRunner:
     def __init__(self, ident):
         self.ident = ident
 
-    async def start(self, cmd):
-        proc = await astart_command([
-            'systemd-cat', '--level-prefix=false', '--identifier='+self.ident,
-            ] + cmd)
+    async def start(self, cmd, *, capture=False):
```

Comment:
So I know I talked to @dbungert about this but I can't remember if I talked to you about this (or if Dan talked to you about this) but I have this scheme I can't work out if it's hair-brained or genius in this area. A problem we can have -- and possibly this is the issue you are seeing with mounts, although I'm not going to claim certainty in this -- is that curtin creates mounts when it runs (to bind mount /dev into the target when running a chrooted process, for example) and if these are either still running or killed too unceremoniously, these mounts can cause problems for other parts of the code.

One way to make this problem just go away would be to use systemd-run to run these commands in a private mount namespace (something like `systemd-run --unit $blah -p PrivateMounts=true python3 -m curtin in-target ....`). Then the mounts that this process creates just don't affect the rest of the system and if we need to terminate it we can just use systemctl stop to make the whole mess go away as if it had never existed.

The output from such a service goes to the journal by default but we can always read that afterwards if we need to look at the output.

One problem this has is that the services run like this have a clean environment -- for most cases we'd probably want some way to propagate the server's environment into the subprocess (I don't really want to try to play the game of deciding which values matter -- very certainly PATH and PYTHONPATH do but which others?). We could do this with a zillion -E flags to systemd-run or by dumping things out to a file and using -p EnvironmentFile=whatever.

What do you think of this idea? Neat or just too complicated?

Reply:
I don't remember discussing this so I'd guess we haven't talked about it.

First impression, this is neat.

A couple years ago, I had a somewhat similar scenario where I was starting a process that would create new routes, IP addresses and whatnot during its lifetime. I banged my head trying to patch it to make sure everything would be cleaned when the process would exit (sometimes unexpectedly)

Turned out that the only reliable way I found was to make that process run in isolation (i.e., in a different network namespace)

I guess this is analogous to what we have here ; except maybe that we have (full) control over curtin.

IMO the downside is that we might be slower to notice issues in curtin when everything gets cleaned up automatically.

Reply:
I'm not sure I would regard curtin not cleaning up after itself when interrupted at arbitrary points as an especially serious bug. Getting that sort of thing 100% right is just a lot of work and we have a mechanism for doing it cleanly, I think we should probably do that.

Reply:
Addressed in https://github.com/canonical/subiquity/pull/1185